### PR TITLE
[WIP] Porting Upstream C++

### PR DESCRIPTION
**Pull request by naftulikay**

### CODE REVIEW

**Comment by naftulikay** [comment]
**EDIT:** Contents [now hosted on the wiki](https://github.com/naftulikay/vw-cdc-emulator/wiki/Upstream-Example).

File: examples/vag_cdc.c [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+uint8_t getCommand(uint32_t cmd);
+void shutdown();
+void softreset();
+
+volatile uint8_t cd;
+volatile uint8_t tr;
+volatile uint8_t mode;
+
+ISR(INT0_vect) //remote signals
```
Comment (naftulikay) [inline examples/vag_cdc.c:64]: Interrupt Service Routine registered for the `INT0` interrupt, presumably registered to tell us when there is data to read from the SPI `MISO` pin for data coming _from_ the head-unit to the CD changer.

#### What it Does

The interrupt service routine for `INT0_vect` detects whether there is data to read from the SPI `MISO` input on the board (reading data from the head unit), and then enters some complicated command capturing logic, setting flags and reading one bit at a time into `cmd`.

File: examples/vag_cdc.c [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+#include <inttypes.h>
+#include <avr/io.h>
+#include <avr/interrupt.h>
+#include <util/delay.h>
+
+#include "uart.h"
+
+#define UART_BAUD_RATE 9600
+#define LED_PWR PA0
+#define RADIO_OUT PD2
```
Comment (naftulikay) [inline examples/vag_cdc.c:19]: Define the `RADIO_OUT` pin/register as being `PD2`.

File: examples/vag_cdc.c [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+#define FT_CBUS1 PD3
+#define RADIO_OUT_IS_HIGH (PIND & (1<<RADIO_OUT))
```
Comment (naftulikay) [inline examples/vag_cdc.c:21]: Let's break this down. 

---

`RADIO_OUT` is the `PD2` register, so we can resolve this into two operations.

```c
1 << PD2
```

The first is a [bitwise left-shift](https://en.wikipedia.org/wiki/Bitwise_operations_in_C#Left_shift_%3C%3C). We shift `1` to the left by the _n_ bits, where _n_ is the current value of the `PD2` register. This has the effect of providing a `1` bit at an arbitrary location, e.g. `1<<0 == 0b0000_0001`, `1<<1=0b0000_0010`, `1<<2=0b0000_0100`. This is basically saying "set bit _n_ to 1."

---

```c
PIND & result
```

The next operation is a [bitwise AND](https://en.wikipedia.org/wiki/Bitwise_operations_in_C#Bitwise_AND_&) between the value of `PIND` and the result of the previous operation. This is essentially saying "check if bit _n_ in `PIND` is set to `1`."

In summary:

> Tell me if the bit number defined by the current value of `PD2` within `PIND` is `1`

In pseudocode, treating it as a list of integers in a bitset (e.g. `PIND=[0, 1, 0, 1, 0, 0, 0, 1]`), each equal to `0` or `1`, instead of a single integer:

```python
PIND[PD2] == 1
```

The result of this operation will inform us as to whether the radio output (presumably `MISO` in SPI for receiving data from the head unit) is set to high, presumably indicating whether

File: examples/vag_cdc.c [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+void shutdown();
+void softreset();
+
+volatile uint8_t cd;
+volatile uint8_t tr;
+volatile uint8_t mode;
+
+ISR(INT0_vect) //remote signals
+{
+	if(RADIO_OUT_IS_HIGH)
```
Comment (naftulikay) [inline examples/vag_cdc.c:66]: As above:

> If there is data ready to be read from `PIND`, which must be the _output_ of the head unit, being the _input_ to the CD changer emulator. For the CD changer, this is `MISO` for receiving data from the head-unit.

File: examples/vag_cdc.c [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+#define MODE_PLAY 0xFF
+#define MODE_SHFFL 0x55
+#define MODE_SCAN 0x00
+
+uint16_t captimehi = 0;
+uint16_t captimelo = 0;
+uint8_t capturingstart = 0;
+uint8_t capturingbytes = 0;
+uint32_t cmd = 0;
```
Comment (naftulikay) [inline examples/vag_cdc.c:51]: The command currently being read from the `MISO` SPI input, being received from the head unit. Each command is 32-bits long.

File: examples/vag_cdc.c [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+		{
+			//uart_puts("bit 0\r\n");
+			cmd = (cmd<<1);
+			cmdbit++;
+		}
+		else
+		{
+			//uart_puts("nothing found\r\n");
+		}
+		if(cmdbit == 32)
```
Comment (naftulikay) [inline examples/vag_cdc.c:99]: If we have captured a full 32-bit command from the head-unit, reset the capturing data so that we treat whatever is next as a new command, and :shipit: ship the previous command as being ready.

File: examples/vag_cdc.c [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+		}
+	}
+	else
+	{
+		captimehi = TCNT1;
+		TCNT1 = 0;
+	}
+}
+
+ISR(INT1_vect) //ft230x cbus1
```
Comment (naftulikay) [inline examples/vag_cdc.c:114]: Interrupt Service Routine for `INT1_vect`, which is a UART [FT230X](https://www.ftdichip.com/Products/ICs/FT230X.html) bridge. This seems to be some kind of initialization code; I think it runs when one of the following possibilities:

 - Event:
    - UART connect?
    - UART disconnect?
 - Init or Event:
    - UART 'boot'?
    - UART connect?

It sets the CD and the track number, which will be written out by the main loop, which reads the `volatile` `cd`, `tr`, and `mode` variables and sends them to the head unit.

I am pretty convinced that this interrupt routine only happens on UART connect or disconnect, or possibly boot or connect. I'll have to look into it more, but it seems that the if statement checks if the `FT_CBUS1` bit within `PIND` is set to `1`, but I'm not sure what that means. It could be checking for UART presence or the type of event.

File: examples/vag_cdc.c [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+		//PORTA &= ~(1<<LED_PWR);
+		cd = 0xBE;
+		tr = 0xFF;
+		mode = 0xFF;
+	}
+	else //usb connect
+	{
+		PORTA |= (1<<LED_PWR);
+		_delay_ms(50);
+		PORTA &= ~(1<<LED_PWR);
```
Comment (naftulikay) [inline examples/vag_cdc.c:127]: `LED_PWR` is an alias for `PA0`. This `else` statement does the following:

 1. Set the `PA0`th bit of `PORTA` without touching anything else.
 2. Wait 50 milliseconds.
 3. Clear the `PA0`th bit of `PORTA` without touching anything else.
 4. Set the volatile `cd` and `tr` variables to some default value.

**TODO:** Why is `LED_PWR` an alias for `PA0`? What is `PORTA`?

**Comment by naftulikay** [comment]
In C, any _non-zero_ value is `true`.

File: examples/vag_cdc.c [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+		cd = 0xB0;
+		tr = 0x00;
+	}
+}
+
+uint8_t spi_xmit(uint8_t val) {
```
Comment (naftulikay) [inline examples/vag_cdc.c:133]: Writes a single byte into `SPDR`, which is presumably the `MOSI` from the Arduino to the head-unit, then spins presumably until the head-unit has confirmed reception of the payload, then returns the current value of `SPDR`, which isn't used anywhere.

The while loop can be expanded like so:

> While the `SPIF`th bit in `SPSR` is _not_ set, spin.

File: examples/vag_cdc.c [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+	SPDR = val;
+	while(!(SPSR & (1<<SPIF)));
+	return SPDR;
+}
+
+void send_package(uint8_t c0, uint8_t c1, uint8_t c2, uint8_t c3, uint8_t c4, uint8_t c5, uint8_t c6, uint8_t c7)
```
Comment (naftulikay) [inline examples/vag_cdc.c:139]: Commands received _from_ the head-unit are four bytes, commands sent _to_ the head-unit are eight bytes.

This sends the data, one byte at a time, sleeping for 874 microseconds between each byte sent. The spin in `spi_xmit` ensures that each byte is received.

File: examples/vag_cdc.c [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+	spi_xmit(c7);
+}
+
+int main(void)
+{
+	cd = 0xBE;
+	tr = 0xFF;
+	mode = 0xFF;
+	//LEDs
+	DDRA |= (1<<LED_PWR);
```
Comment (naftulikay) [inline examples/vag_cdc.c:164]: An LED? Sets the `PA0`th bit of `DDRA` to `1`.

File: examples/vag_cdc.c [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+
+	//pullup
+	PORTD |= (1<<FT_CBUS1);
```
Comment (naftulikay) [inline examples/vag_cdc.c:167]: This is the only time that `PORTD` is referenced. What is it? In any case, this sets the `FT_CBUS1`th bit of `PORTD` to `1`. 

`FT_CBUS1` is referenced above as being related to UART.

File: examples/vag_cdc.c [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+
+	uart_init(UART_BAUD_SELECT(UART_BAUD_RATE, F_CPU));
```
Comment (naftulikay) [inline examples/vag_cdc.c:169]: Initialize the UART device, setting the baud rate, and [`UART_BAUD_SELECT`, defined in `uart/uart.h`](https://github.com/naftulikay/avr/blob/c86fe2621d07623187ff82ed44a9e19198c0ff51/uart/uart.h#L116) sets the baud rate according to the CPU frequency:

```c
/** @brief  UART Baudrate Expression
 *  @param  xtalcpu  system clock in Mhz, e.g. 4000000L for 4Mhz          
 *  @param  baudrate baudrate in bps, e.g. 1200, 2400, 9600     
 */
#define UART_BAUD_SELECT(baud_rate, xtal_cpu) \
  ((xtal_cpu) / ((baud_rate) * 16l) - 1)
```

File: examples/vag_cdc.c [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+
+	//init SPI
+	DDRB |= (1<<PB5) | (1<<PB7)| (1<<DDB4);
```
Comment (naftulikay) [inline examples/vag_cdc.c:172]: Set the `PB5`th, `PB7`th, and `DDB4`th bits of `DDRB` to `1`.

```python
DDRB[PB5], DDRB[PB7], DDRB[DDB4] = 1, 1, 1
```

Judging by the comments, it's assumed that these pins are related to SPI, at the very least `DDRB`.

File: examples/vag_cdc.c [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+
+	// SPI Type: Master
+	// SPI Clock Rate: 62,500 kHz
+	// SPI Clock Phase: Cycle Start
+	// SPI Clock Polarity: Low
+	// SPI Data Order: MSB First
+	SPCR=0x57;//at 8MHz
```
Comment (naftulikay) [inline examples/vag_cdc.c:179]: Generic SPI information to help clue us in on the way things work.

 - Type: Master
 - Clock Rate: 62,500 kHz
 - Clock Phase: Cycle Start
 - Clock Polarity: Low
 - Data Order: Most Significant Bit First

`SPCR` seems to be "SPI Clock Rate," and `0x57` is `87` in decimal. `0x56`, referenced below (for use "at 4MHz") is `86` in decimal. I'm not sure how exactly this correlates: wouldn't the clock rate be related by a factor of two at half the hertz? Instead, it's only different by `1`.

File: examples/vag_cdc.c [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+	//SPCR=0x56;//at 4MHz
+	SPSR=0x00;
```
Comment (naftulikay) [inline examples/vag_cdc.c:181]: This is the SPI Status Register, according to the documentation. Pulling this low probably has to do with the clock polarity above.

File: examples/vag_cdc.c [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+					_delay_ms(100);
+					shutdown();
+					PORTA |= (1<<LED_PWR);
+				}*/
+			}
+		}
+    }
+}
+
+uint8_t getCommand(uint32_t cmd)
```
Comment (naftulikay) [inline examples/vag_cdc.c:254]: Read the volatile `cmd` variable, which is 4 bytes of binary data.

File: examples/vag_cdc.c [hunk 18]
```diff
@@ -0,0 +1,260 @@
...
+		}
+		//                disc  trk  min  sec
+		//send_package(0x34,cd,tr,0xFF,0xFF,0xFF,0xCF,0x3C);
+		send_package(0x34,cd,tr,0xFF,0xFF,mode,0xCF,0x3C);
+        _delay_ms(41);
+		if (newcmd)
+		{
+			newcmd = 0;
+			uint8_t c = getCommand(cmd);
+			if (c)
```
Comment (naftulikay) [inline examples/vag_cdc.c:234]: If there is a command (non-zero), write it to UART.

File: examples/vag_cdc.c [hunk 19]
```diff
@@ -0,0 +1,260 @@
...
+	uart_putc(0xAA);
+	uart_putc(0x55);
+	PORTA &= ~(1<<LED_PWR);
+
+	send_package(0x74,0xBE,0xFE,0xFF,0xFF,0xFF,0x8F,0x7C); //idle
+	_delay_ms(10);
+	send_package(0x34,0xFF,0xFE,0xFE,0xFE,0xFF,0xFA,0x3C); //load disc
+	_delay_ms(100);
+	send_package(0x74,0xBE,0xFE,0xFF,0xFF,0xFF,0x8F,0x7C); //idle
+	_delay_ms(10);
```
Comment (naftulikay) [inline examples/vag_cdc.c:202]: These three commands tell the head unit "HELLO I AM CD CHANGER U CAN TRUST ME."

File: examples/vag_cdc.c [hunk 20]
```diff
@@ -0,0 +1,260 @@
...
+	// SPI Data Order: MSB First
+	SPCR=0x57;//at 8MHz
+	//SPCR=0x56;//at 4MHz
+	SPSR=0x00;
+
+	//beta commands -> cdc
+	TCCR1B |= (1<<CS11);    // no prescaler 8 -> 1 timer clock tick is 1us long
+	GICR |= (1<<INT0) | (1<<INT1);
+	MCUCR |= (1<<ISC00) | (1<<ISC10); //any change on INT0 and INT1
+	sei();
```
Comment (naftulikay) [inline examples/vag_cdc.c:187]: As defined in `avr/interrupt.h`, the `sei` command _enables_ interrupts, which appear disabled by default. See: `man sei`.

File: examples/vag_cdc.c [hunk 21]
```diff
@@ -0,0 +1,260 @@
...
+					cd = 0xFF^(r & 0x0F);
+			}
+			//send inverted TR No.
+			else
+				tr = 0xFF^r;
+		}
+		//                disc  trk  min  sec
+		//send_package(0x34,cd,tr,0xFF,0xFF,0xFF,0xCF,0x3C);
+		send_package(0x34,cd,tr,0xFF,0xFF,mode,0xCF,0x3C);
+        _delay_ms(41);
```
Comment (naftulikay) [inline examples/vag_cdc.c:229]: After receiving a command from UART, send the package to the head-unit, then delay 41 milliseconds, then try to read a command if it's available and send it back to UART.

File: examples/vag_cdc.c [hunk 22]
```diff
@@ -0,0 +1,260 @@
+/*
+ * VAG_CDC.c
+ *
+ * Created: 23.06.2013 20:00:51
+ *  Author: Dennis Schuett, dev.shyd.de
+ */
+
+#define F_CPU 8000000UL
```
Comment (naftulikay) [inline examples/vag_cdc.c:8]: Define CPU frequency.

File: examples/vag_cdc.c [hunk 23]
```diff
@@ -0,0 +1,260 @@
...
+	// SPI Clock Phase: Cycle Start
+	// SPI Clock Polarity: Low
+	// SPI Data Order: MSB First
+	SPCR=0x57;//at 8MHz
+	//SPCR=0x56;//at 4MHz
+	SPSR=0x00;
+
+	//beta commands -> cdc
+	TCCR1B |= (1<<CS11);    // no prescaler 8 -> 1 timer clock tick is 1us long
+	GICR |= (1<<INT0) | (1<<INT1);
```
Comment (naftulikay) [inline examples/vag_cdc.c:185]: Register interrupt 0 (`INT0`) and interrupt 1 (`INT1`) with the global interrupt control register (`GICR`).

File: examples/vag_cdc.c [hunk 24]
```diff
@@ -0,0 +1,260 @@
...
+	MCUCR |= (1<<ISC00) | (1<<ISC10); //any change on INT0 and INT1
```
Comment (naftulikay) [inline examples/vag_cdc.c:186]: `MCUCR` is defined in the docs as:

> MCU general Control Register

[This is related to external interrupts as documented here](http://www.avr-tutorials.com/interrupts/The-AVR-8-Bits-Microcontrollers-External-Interrupts). I think according to those docs, writing to `MCUCR` tells it which events to listen for, of:

 - Low Level
 - Any Logic Change
 - Falling Edge
 - Rising Edge

And:

 - `ISC00` is defined as: Any Logic Change + Rising Edge for `INT0`.
 - `ISC10` is defined as: Any Logic Change + Rising Edge for `INT1`.

File: examples/vag_cdc.c [hunk 25]
```diff
@@ -0,0 +1,260 @@
...
+volatile uint8_t tr;
+volatile uint8_t mode;
+
+ISR(INT0_vect) //remote signals
+{
+	if(RADIO_OUT_IS_HIGH)
+	{
+		if (capturingstart || capturingbytes)
+		{
+			captimelo = TCNT1;
```
Comment (naftulikay) [inline examples/vag_cdc.c:70]: [`TCNT` is a clock/timer count](http://www.avrbeginners.net/architecture/timers/timers.html), which is used here to establish spans of time for capturing data.

File: examples/vag_cdc.c [hunk 26]
```diff
@@ -0,0 +1,260 @@
...
+	// SPI Clock Rate: 62,500 kHz
+	// SPI Clock Phase: Cycle Start
+	// SPI Clock Polarity: Low
+	// SPI Data Order: MSB First
+	SPCR=0x57;//at 8MHz
+	//SPCR=0x56;//at 4MHz
+	SPSR=0x00;
+
+	//beta commands -> cdc
+	TCCR1B |= (1<<CS11);    // no prescaler 8 -> 1 timer clock tick is 1us long
```
Comment (naftulikay) [inline examples/vag_cdc.c:184]: [`TCCR1B`](http://www.avrbeginners.net/architecture/timers/timers.html) is an AVR timer. `CS11` is a "clock select" bit, which basically changes the clock rate on the board, with the definition of:

 - Divide Clock by 8
 - Divide Clock by 64
 - Increment Timer 1 on T1 Pin Falling Edge
 - Increment Timer 1 on T1 Pin Rising Edge

`TCNT1` counts system clock ticks according to the frequency specified.
